tests/zfs-tests/tests/functional/zvol/zvol_misc/zvol_misc_volmode.ksh

```diff
@@ -0,0 +1,287 @@
+	# that can affect device nodes
+	for i in {1..3}; do
+		udev_wait
+		[[ ! -e "$device" ]] && return 0
+	done
+	log_fail "$device exists when not expected"
+}
+
+#
+# Verify $property on $dataset is inherited by $parent and is set to $value
+#
+function verify_inherited # property value dataset parent
+{
+	typeset property="$1"
+	typeset value="$2"
+	typeset dataset="$3"
+	typeset parent="$4"
+
+	typeset val=$(get_prop "$property" "$dataset")
+	typeset src=$(get_source "$property" "$dataset")
+	if [[ "$val" != "$value" || "$src" != "inherited from $parent" ]]
+	then
+		log_fail "Dataset $dataset did not inherit $property properly:"\
+		    "expected=$value, value=$val, source=$src."
+	fi
+}
+
+#
+# Create a small partition on $device, then verify if we can access it
+#
+function verify_partition # device
+{
+	typeset device="$1"
+
+	if [[ ! -b "$device" ]]; then
+		log_fail "$device is not a block device"
+	fi
+	# create a small dummy partition: unfortunately we can't use
+	# set_partition() here, becase it internally calls log_must()
+	if is_linux; then
+		parted $device -s -- mklabel gpt || return $?
+		parted $device -s -- mkpart part0 1cyl 1cyl || return $?
+		blockdev --rereadpt $device 2>/dev/null
+	else
+		typeset format_file="$TESTDIR/format_in.$$"
+		echo "partition" >$format_file
+                echo "0" >> $format_file
+                echo "" >> $format_file
+                echo "" >> $format_file
+                echo "1" >> $format_file
+                echo "1m" >> $format_file
+                echo "label" >> $format_file
+                echo "" >> $format_file
+                echo "q" >> $format_file
+                echo "q" >> $format_file
+                format -e -s -d $device -f "$format_file"
+	fi
+	# verify we can access the partition on the device
+	devname="$(readlink -f "$device")"
+	if is_linux; then
+		[[ -b "$devname""p1" ]]
+	else
+		[[ -b "$devname""s0" ]]
+	fi
+	return $?
+}
+
+#
+# Set volmode tunable to $value
+#
+function sysctl_volmode # value
+{
+	typeset value="$1"
+
+	log_note "Setting volmode tunable to $value"
+	if is_linux; then
+		echo "$value" > '/sys/module/zfs/parameters/zvol_volmode'
+	else
+		sysctl 'vfs.zfs.vol.mode' "$value"
+	fi
+	if [[ $? -ne 0 ]]; then
+		log_fail "Unable to set volmode tunable to $value"
+	fi
+}
+
+#
+# Wait for udev to settle, completely.
+# This is quite discomforting, but there's a race condition here
+# (Amazon 2015.09 x86_64 Release (TEST) is good at triggering this)  where the
+# kernel tries to remove zvol device nodes while they're open by [blkid],
+# [zvol_id] or other udev related processes.
+#
+function udev_wait
+{
+	log_must sleep 1
+	udevadm trigger --action=change
+	udevadm settle
+	for i in {1..3}; do
+		blkid="$(pgrep blkid | wc -l)"
+		zvol_id="$(pgrep zvol_id | wc -l)"
+		[[ "0" == "$zvol_id" && "0" == "$blkid" ]] && return
+		udevadm settle
+	done
+	log_fail "Wait timeout reached for udev_wait"
+}
```

Comment:
Let's move all of these helper functions which are also needed by `zvol_misc_snapdev.ksh` to `tests/zfs-tests/tests/functional/zvol/zvol_misc/zvol_misc.ksh` and source it for both test cases.